apps/spark/src/main/java/com/linkedin/openhouse/jobs/spark/Operations.java

```
@@ -270,10 +279,72 @@ public static ChronoUnit convertGranularityToChrono(String granularity) {
+    Expression filter =
+        SparkJobUtil.createDeleteFilter(columnName, columnPattern, granularity, count);
+    TableScan scan = table.newScan().filter(filter);
+    try (CloseableIterable<FileScanTask> filesIterable = scan.planFiles()) {
+      List<FileScanTask> filesList = Lists.newArrayList(filesIterable);
+      return filesList.stream()
+          .collect(
+              Collectors.groupingBy(
+                  task -> {
+                    String path = task.file().path().toString();
+                    return Paths.get(path).getParent().toString();
+                  },
+                  Collectors.mapping(task -> task.file().path().toString(), Collectors.toList())));
+    } catch (IOException e) {
+      throw new RuntimeIOException(e, "Failed to close table scan: %s", scan);
+    }
+  }
+
+  private void writeBackupDataManifests(
+      Map<String, List<String>> manifestCache, Table table, String backupDir) {
+    try {
+      final FileSystem fs = fs();
+      manifestCache.forEach(
+          (partitionPath, files) -> {
+            // Rename data files to backup dir
+            List<String> backupFiles =
+                files.stream()
+                    .map(file -> getTrashPath(table, file, backupDir).toString())
+                    .collect(Collectors.toList());
+            Map<String, Object> jsonMap = new HashMap<>();
+            jsonMap.put("file_count", backupFiles.size());
+            jsonMap.put("files", backupFiles);
+            String jsonStr = new Gson().toJson(jsonMap);
+            // Create data_manifest file
+            Path destPath = new Path(partitionPath, "data_manifest.json");
+            try (FSDataOutputStream out = fs.create(destPath, true)) {
+              out.write(jsonStr.getBytes());
+              out.flush();
+              log.info("Wrote {} with {} files", destPath, files.size());
+            } catch (IOException e) {
+              log.error("Failed to write manifest {}", destPath, e);
```

raise exception, let the app fail